Add `mock_call` (#385)
foundry-rs/starknet-foundry: docs/src/appendix/cheatcodes/start_mock_call.md

# `start_mock_call`

> `fn start_mock_call(contract_address: ContractAddress, function_name: felt252, ret_data: Array::<felt252>)`

> Mocks external function `function_name` for a contract at the given address. The function `function_name` will return data provided in `ret_data` argument.

Author: @MaksymilianDemitraszek Maybe we should mention that only external functions can be mocked?

> Mocks external function `function_name` for a contract at the given address. The function `function_name` will return data provided in `ret_data` argument.

Reviewer: I think we should use phrasing `Mocks contract call` instead of `Mocks function`

> Mocks external function `function_name` for a contract at the given address. The function `function_name` will return data provided in `ret_data` argument.

Author: Changed in https://github.com/foundry-rs/starknet-foundry/pull/385/commits/8df3ebecb7e30ae022af1774dd025188108337f3

This change can be canceled with [`stop_mock_call`](./stop_mock_call.md).

- `contract_address` - target contract address

> - `fn_name` - name of the function in contract `contract_address` that will be mocked

Reviewer 2: ```suggestion
- `fn_name` - name of the function in a contract at the `contract_address` that will be mocked
```

# `start_mock_call`

> `fn start_mock_call(contract_address: ContractAddress, function_name: felt252, ret_data: Array::<felt252>)`

> Mocks contract call to a `function_name` of a contract at the given address. The function `function_name` will return data provided in `ret_data` argument. 

Reviewer: ```suggestion
Mocks contract call to a `function_name` of a contract at the given address. A call to function `function_name` will return data provided in `ret_data` argument. 
```

# `start_mock_call`

> `fn start_mock_call<T, impl TSerde: serde::Serde<T>, impl TDestruct: Destruct<T>>(contract_address: ContractAddress, function_name: felt252, ret_data: T)`

Mocks contract call to a `function_name` of a contract at the given address. A call to function `function_name` will return data provided in `ret_data` argument. 

> If contract is deployed at given address, mocked function won't be executed. Address with no contract can be mocked as well.

Reviewer 3: `If contract is deployed at given address, mocked function won't be executed`
Wdym? I don't think this sentence is particularly helpful, it confused me at first

Author: There could be no contract deployed under the given address. If there is, and the call to the function was mocked, then this function will not be executed.

Author: Slightly updated in https://github.com/foundry-rs/starknet-foundry/pull/385/commits/dc2461084f0a10b3d556f4b6f51b177251552516

> Mocks contract call to a `function_name` of a contract at the given address. A call to function `function_name` will return data provided in `ret_data` argument. 
> 
> If contract is deployed at given address, mocked function won't be executed. Address with no contract can be mocked as well.

Reviewer 3: ```suggestion
Mocks contract call to a `function_name` of a contract at the given address. A call to function `function_name` will return data provided in `ret_data` argument. Addresses of non-existing contracts can be mocked as well.
```

Mock can be canceled with [`stop_mock_call`](./stop_mock_call.md).

- `contract_address` - target contract address
- `function_name` - name of the function in a contract at the `contract_address` that will be mocked
- `ret_data` - data to return by the function `function_name`

> 📝 **Note**

> > The inner call (i.e. when a contract calls a function from within itself) cannot be mocked.

Reviewer 3: ```suggestion
> The inner call  - a call when a contract calls a function from within itself - cannot be mocked.
```

Author: I don't think this version is more readable. And the word `call` repeats too many times.

Reviewer 3: Agree, can we remove the `i. e.` though? I think it is the only case of an inner call, no?

Author: It is used here to indicate clarification

Reviewer 3: Thanks, mb, I was sure was en equivalent of giving an example!

For contract implementation:

```rust
// ...
#[external(v0)]
impl IContractImpl of IContract<ContractState> {
    #[storage]
    struct Storage {
        // ...
        
        balance: felt252
    }

    fn set_balance(ref self: ContractState, new_balance: felt252) {
        self.balance.write(new_balance);
    }

    fn get_balance(self: @ContractState) -> felt252 {
        self.balance.read()
    }
}
// ...
```

We can use `start_mock_call` in a test to change the data returned by `get_balance` for a given contract:

```rust
#[test]
fn test_mock_call() {
    // ...
    
    let mock_ret_data = 421;

>     start_mock_call(contract_address, 'get_balance', mock_ret_data);

Reviewer 3: outdated signature

Author: Why is this outdated? `start_mock_call` is called the same way in tests.
